### PR TITLE
Clean up l_object.cpp

### PR DESCRIPTION
**Pull request by Zughy**
So, I've decided cleaning up MT code on the go. I think a consistent ordered code is vital both for core devs and for possible contributors. This is, in short, what I did:

* consistent names
  * `ObjectRef = obj`
  * `ServerActiveObject = sao`
  * `LuaEntitySAO = entitysao`
  * `PlayerSAO = playersao`
  * `RemotePlayer = player`
* consistent structure: check for object, variables declaration, execution+return. These sections are separated by an empty line. When functions were too long, the execution got split up accordingly.
* removed all `NULL` checks. Now there are `nullptr`s instead, as I find it more straightforward than `!something`
* removed useless comments, e.g. `//Do it`
* removed `is_player_connected`, a deprecated method that wasn't even listed in the API
* descriptions above functions now match the API
* removed info about parameters in the functions description: they're already explained in the API
* edited the API where needed
* simplified a few methods to make them more readable e.g. `set_bone_position`
* added missing Lua checks ( partially solves #1794 )

Things I'm not sure about, core dev help needed:
* ~how to name `p` in `set_sprite(...)`~
* the old `-1` value of HUD IDs, returned when ID was not a number. I had a look at the code and it doesn't look to me was meaningful at all so I simply removed it, but a double check is always better
* ~debug lines at 2249 and 2257. Do you want to keep them?~
* TODO at 483: has it been solved?

## To do

This PR is Ready for Review.

## How to test

Well: check the code. Also, I might have made a few mistakes as I'm human, even if I double-checked my edits twice. So idk, take a solid game, play it a little bit and see if it all works accordingly.

Lastly, I want to underline how, imho, this should be high priority: I've seen things which were pretty messed up, like edits pushed a couple of weeks ago going completely against the guidelines (the minimap), or useless 8 year old comments who nobody apparently decided to remove. A tidy environment helps a lot. And I've actually had fun doing it, it was relaxing

### CODE REVIEW

File: doc/lua_api.txt [hunk 1]
```diff
@@ -6420,8 +6420,8 @@ object you are working with still exists.
           nodes square)
     * `selected_mode` is the mode index to be selected after modes have been changed
     (0 is the first mode).
-* `set_sky(parameters)`
-    * `parameters` is a table with the following optional fields:
+* `set_sky(skybox_parameters)`
```
Comment (paramat) [inline doc/lua_api.txt:6423]: 'parameters' is better because these parameters contain several things that are unrelated to 'skybox'. Also because 'skybox' is used in the parameters to mean one particular sky mode.

Comment (paramat) [reply doc/lua_api.txt:6423]: Perhaps 'sky_parameters'?

**Comment by paramat** [comment]
> how to name p in set_sprite(...)

> `set_sprite(p, num_frames, framelength, select_x_by_camera)`
`p`: {x=column number, y=row number}, the coordinate of the first frame default: `{x=0, y=0}`

How about 'first_frame' or 'start_frame'?

File: doc/lua_api.txt (hunk 1 above)
Comment (paramat) [reply doc/lua_api.txt:6423]: Ah, i see we have:
`SkyboxParams skybox_params = player->getSkyParams();`
in code, which might be why you chose this, but that line also has `getSkyParams`, so 'sky_parameters' is suitable.

File: src/script/lua_api/l_object.cpp [hunk 2]
```diff
@@ -1717,26 +1690,27 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
...
 		return 0;
 
 	const std::string &name = player->getHotbarSelectedImage();
+
 	lua_pushlstring(L, name.c_str(), name.size());
 	return 1;
 }
 
-// set_sky(self, {base_color=, type=, textures=, clouds=, sky_colors={}})
+// set_sky(self, skybox_parameters)
```
Comment (paramat) [inline src/script/lua_api/l_object.cpp:1702]: sky_parameters

File: doc/lua_api.txt (hunk 1 above)
Comment (paramat) [reply doc/lua_api.txt:6423]: To be clear, i am not suggesting altering `skybox_params` in the code, i am suggesting using `sky_parameters` in docs and code comments only.

**Comment by Zughy** [comment]
as a non native speaker, `start_frame` is more intuitive to me. I'll wait for more feedback

**Comment by sfan5** [comment]
> how to name p in set_sprite(...)

`frame_pos` or `sprite_pos`

> debug lines at 2249 and 2257. Do you want to keep them?

No, anyone can write them again if they need them.

　

Regarding the renaming: I think it's useful to change the variable if it's misnamed (e.g. `ServerActiveObject *co`), but changing 
`o` to `obj` in a short function is not an improvement.
Also I don't think `!player` to `player == nullptr` is an improvement.

File: doc/lua_api.txt [hunk 3]
```diff
@@ -6494,21 +6494,17 @@ object you are working with still exists.
...
+* `set_local_animation(idle, walk, dig, walk_while_dig, frame_speed)`:
+  set animation for player model in third person view.
+    * Every animation equals to a `{x=starting frame, y=ending frame}` table.
+    * `frame_speed` sets the animations frame speed. Default is 30.
+* `get_local_animation()`: returns idle, walk, dig, dig+walk tables and
   `frame_speed`.
-* `set_eye_offset({x=0,y=0,z=0},{x=0,y=0,z=0})`: defines offset value for
-  camera per player.
+* `set_eye_offset(firstperson, thirdperson)`: defines offset value
+  for camera per player, as a table.
```
Comment (sfan5) [inline doc/lua_api.txt:6504]: I don't know what's with the weird need to mention tables everywhere.
If something is a position, it should be obvious that it's a `{x=0, y=0, z=0}` table:
https://github.com/minetest/minetest/blob/738f62421872c0be5fcd483ad4573e5d879e7418/doc/lua_api.txt#L1469-L1477

**Comment by Zughy** [comment]
> Also I don't think !player to player == nullptr is an improvement.

It was simply a matter of choosing one or the other, as the code featured every variation of it (a lot of them also missed the new line for the body). Relevant discussion: https://github.com/minetest/minetest/pull/10480#discussion_r503529833. As someone who's quite a beginner in C++, `nullptr` helps me understand better all the pointers thing

File: doc/lua_api.txt [hunk 4]
```diff
@@ -6494,21 +6494,17 @@ object you are working with still exists.
...
+* `get_local_animation()`: returns idle, walk, dig, dig+walk tables and
   `frame_speed`.
-* `set_eye_offset({x=0,y=0,z=0},{x=0,y=0,z=0})`: defines offset value for
-  camera per player.
+* `set_eye_offset(firstperson, thirdperson)`: defines offset value for camera
+  per player.
     * in first person view
     * in third person view (max. values `{x=-10/10,y=-10,15,z=-5/5}`)
-* `get_eye_offset()`: returns `offset_first` and `offset_third`
+* `get_eye_offset()`: returns first and third person offsets as tables.
```
Comment (sfan5) [inline doc/lua_api.txt:6507]: > tables

**Comment by sfan5** [comment]
> It was simply a matter of choosing one or the other, as the code featured every variation of it

My point here is that changing bad style (`x == NULL`) to good style (e.g. `x == nullptr`) is useful, but changing good style (`!x`) to a different good style is not useful.

> As someone who's quite a beginner in C++, nullptr helps me understand better all the pointers thing, that's why I opted for it

...just like this wouldn't justify changing `!foo` to `foo == nullptr` everywhere in Minetest.

File: doc/lua_api.txt (hunk 1 above)
Comment (paramat) [reply doc/lua_api.txt]: Done, thanks.

File: src/script/lua_api/l_object.cpp (hunk 2 above)
Comment (paramat) [reply src/script/lua_api/l_object.cpp]: Done.

**Comment by Zughy** [comment]
I understand your point now, I'll avoid doing that in the future, I'm sorry (but please don't make me revert the ones I've already edited, it's... a lot)

**Comment by v-rob** [comment]
Reverting shouldn't be necessary at this point; it's not like it's a very hard change to review, even if there are a lot.

File: src/script/lua_api/l_object.cpp [hunk 5]
```diff
@@ -606,39 +530,34 @@ int ObjectRef::l_set_animation_frame_speed(lua_State *L)
...
-	if (!lua_isnil(L, 4))
-		rotation = check_v3f(L, 4);
-	co->setBonePosition(bone, position, rotation);
+	ServerActiveObject *sao = getobject(ref);
+	if (sao == nullptr)
+		return 0;
+
+	std::string bone = readParam<std::string>(L, 2);
+	v3f position = !lua_isnil(L, 3) ? check_v3f(L, 3) : v3f(0, 0, 0);
+	v3f rotation = !lua_isnil(L, 4) ? check_v3f(L, 4) : v3f(0, 0, 0);
```
Comment (v-rob) [inline src/script/lua_api/l_object.cpp:558]: In this case, why not just `v3f position = read_v3f(L, 3);`?  If it's nil, it will automatically be zeroed.  Same in many other places.

File: src/script/lua_api/l_object.cpp [hunk 6]
```diff
@@ -668,68 +585,57 @@ int ObjectRef::l_get_bone_position(lua_State *L)
...
-	bone = "";
-	if (!lua_isnil(L, 3))
-		bone = readParam<std::string>(L, 3);
-	position = v3f(0, 0, 0);
-	if (!lua_isnil(L, 4))
-		position = read_v3f(L, 4);
-	rotation = v3f(0, 0, 0);
-	if (!lua_isnil(L, 5))
-		rotation = read_v3f(L, 5);
+	bone      = lua_isnil(L, 3) ? ""           : readParam<std::string>(L, 3);
```
Comment (v-rob) [inline src/script/lua_api/l_object.cpp:609]: How about `bone = readParam<std::string>(L, 3, "")` for a default return value?

Sometime in the future (not in this PR), `readParam` should be made comprehensive and work for all sorts of types, e.g. s32, v3s16, and suchlike, all with default values options.  That way, we don't have to have such a mixture of `readParam` and `luaL_check<type>` and `read_<type>` with `nil` and type checks all over the place.

File: src/script/lua_api/l_object.cpp [hunk 7]
```diff
@@ -796,12 +701,14 @@ int ObjectRef::l_get_properties(lua_State *L)
...
 	ObjectRef *ref = checkobject(L, 1);
-	ServerActiveObject *co = getobject(ref);
-	if (co == NULL)
+	ServerActiveObject *sao = getobject(ref);
+	if (sao == nullptr)
 		return 0;
-	ObjectProperties *prop = co->accessObjectProperties();
+
+	ObjectProperties *prop = sao->accessObjectProperties();
 	if (!prop)
```
Comment (v-rob) [inline src/script/lua_api/l_object.cpp:709]: Since everything else is a `nullptr` now, I suppose it wouldn't hurt to make it consistent and have one more here that you missed.  But it's not important.

File: src/script/lua_api/l_object.cpp [hunk 8]
```diff
@@ -950,59 +859,61 @@ int ObjectRef::l_get_acceleration(lua_State *L)
...
+		return 0;
 
-	if (co == NULL) return 0;
 	if (isNaN(L, 2))
 		throw LuaError("ObjectRef::set_yaw: NaN value is not allowed.");
 
 	float yaw = readParam<float>(L, 2) * core::RADTODEG;
-	co->setRotation(v3f(0, yaw, 0));
+
+	entitysao->setRotation(v3f(0, yaw, 0));
```
Comment (v-rob) [inline src/script/lua_api/l_object.cpp:916]: Question (not related to this cleanup, just a general question), but doesn't setting the other two to `0` reset the pitch and roll of `ObjectRef:set_rotation`?  If that's so, sounds like a bug.

**Review by v-rob — COMMENTED** [review]
Looks pretty good.  I don't like the weird type checks everywhere, but that's better solved with better Lua-interfacing functions, like extending `readParam` and adding something like `paramIsType`.  But that's for another PR, so I won't press inconsistent type checking here.

File: src/script/lua_api/l_object.cpp (hunk 5 above)
Comment (Zughy) [reply src/script/lua_api/l_object.cpp]: Nice catch, done :)

File: src/script/lua_api/l_object.cpp (hunk 6 above)
Comment (Zughy) [reply src/script/lua_api/l_object.cpp]: nice idea about the standard

**Comment by Zughy** [comment]
Thanks v-rob for the in-depth review, all problems have been addressed, and `p` is now `start_frame`. I'd like to remind the other 2 doubts of mine:
* the old -1 value of HUD IDs, returned when ID was not a number. I had a look at the code and it doesn't look to me was meaningful at all so I simply removed it, but a double check is always better
* TODO at 483: has it been solved?

Also, I noticed readParam<bool> and <std::string> take an optional default value, while others don't. Is it normal? (helper.cpp)

File: doc/lua_api.txt (hunk 3 above)
Comment (paramat) [reply doc/lua_api.txt]: In fact, mentioning 'table' is confusing, these are positions. Positions are technically tables but are not thought of that way, they are thought of as positions.

File: src/script/lua_api/l_object.cpp (hunk 8 above)
Comment (rubenwardy) [reply src/script/lua_api/l_object.cpp:910]: I have a feeling that doing this any other way may mess with XYZ/ZYX ordering - ie: the yaw won't turn out correctly

**Review by rubenwardy — APPROVED** [review]

File: doc/lua_api.txt [hunk 9]
```diff
@@ -6494,21 +6494,17 @@ object you are working with still exists.
...
-* `get_local_animation()`: returns stand, walk, dig, dig+walk tables and
+* `set_local_animation(idle, walk, dig, walk_while_dig, frame_speed)`:
+  set animation for player model in third person view.
+    * Every animation equals to a `{x=starting frame, y=ending frame}` table.
+    * `frame_speed` sets the animations frame speed. Default is 30.
+* `get_local_animation()`: returns idle, walk, dig, walk_while_dig tables and
   `frame_speed`.
-* `set_eye_offset({x=0,y=0,z=0},{x=0,y=0,z=0})`: defines offset value for
-  camera per player.
+* `set_eye_offset(firstperson, thirdperson)`: defines offset value for camera
```
Comment (paramat) [inline doc/lua_api.txt:6503]: Would be more correct to call these `offset vectors`, they are not values and there are 2 of them.

**Review by v-rob — APPROVED** [review]
Looks good.  I'm glad to see cleanups and code consistency.

**Comment by Lejo1** [comment]
As I just recolonized this PR removes the pretty well used player:is_player_connected function and breaks some mods on the way.
What's the replace for it? Is it minetest.is_player?
If yes I think it should be clear that it only returns true if the player is online.

**Comment by sfan5** [comment]
`is_player_connected` has been marked deprecated since 5.2.0 (released 9 months ago) **and** there has been a deprecation warning the whole time.
Where exactly is it well used?

**Comment by Lejo1** [comment]
>  Where exactly is it well used?

That might sound silly. It's just I always used this functions like any other so I expected to be official. I learned minetest modding mostly by doing so I must have seen the function somewhere else.
Anyhow such changes are needed. Is `minetest.is_player` now the official replacement?

**Comment by sfan5** [comment]
> Is `minetest.is_player` now the official replacement?

is_player does not do the same thing.
If you absolutely have to determine if an ObjectRef is valid you can use e.g. `get_yaw() ~= nil`, but you should really rework your code to pass along the player name to `minetest.after` and then retrieve the player again.

> Or is it just completely removed and my maybe dirty way of storing it using minetst.after and using it later is not anymore supported?

See here: (section was added with 5.2.0)
https://github.com/minetest/minetest/blob/68cd93b8657c0c8cad0196e79fb63e3275322e38/doc/lua_api.txt#L6107-L6118

File: src/script/lua_api/l_object.cpp [hunk 10]
```diff
@@ -541,27 +464,22 @@ int ObjectRef::l_get_local_animation(lua_State *L)
...
-	// Do it
-	v3f offset_first = v3f(0, 0, 0);
-	v3f offset_third = v3f(0, 0, 0);
 
-	if (!lua_isnil(L, 2))
-		offset_first = read_v3f(L, 2);
-	if (!lua_isnil(L, 3))
-		offset_third = read_v3f(L, 3);
+	v3f offset_first = read_v3f(L, 2);
+	v3f offset_third = read_v3f(L, 3);
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp:477]: These are now no longer optional. `nil` values will result in errors.

File: src/script/lua_api/l_object.cpp [hunk 11]
```diff
@@ -606,39 +524,34 @@ int ObjectRef::l_set_animation_frame_speed(lua_State *L)
...
-	if (!lua_isnil(L, 4))
-		rotation = check_v3f(L, 4);
-	co->setBonePosition(bone, position, rotation);
+	ServerActiveObject *sao = getobject(ref);
+	if (sao == nullptr)
+		return 0;
+
+	std::string bone = readParam<std::string>(L, 2);
+	v3f position = check_v3f(L, 3);
+	v3f rotation = check_v3f(L, 4);
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp:552]: Arguments are no longer optional. `nil` will cause errors.

File: src/script/lua_api/l_object.cpp [hunk 12]
```diff
@@ -668,68 +579,57 @@ int ObjectRef::l_get_bone_position(lua_State *L)
...
-		bone = readParam<std::string>(L, 3);
-	position = v3f(0, 0, 0);
-	if (!lua_isnil(L, 4))
-		position = read_v3f(L, 4);
-	rotation = v3f(0, 0, 0);
-	if (!lua_isnil(L, 5))
-		rotation = read_v3f(L, 5);
+	bone      = readParam<std::string>(L, 3, "");
+	position  = read_v3f(L, 4);
+	rotation  = read_v3f(L, 5);
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp:605]: Same here, for pos and rot.

File: src/script/lua_api/l_object.cpp [hunk 13]
```diff
@@ -2220,26 +2184,18 @@ int ObjectRef::l_get_day_night_ratio(lua_State *L)
...
-	if (!lua_istable(L, 2)) {
-		return 0;
-	}
-
-	// Arg 3 should be a number
-	s16 wantedmode = lua_tonumber(L, 3);
-
+	luaL_checktype(L, 2, LUA_TTABLE);
 	std::vector<MinimapMode> modes;
+	s16 selected_mode = luaL_checkint(L, 3);
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp:2198]: `"2"` is a valid number but fails in `luaL_checkint`.

File: src/script/lua_api/l_object.cpp (hunk 11 above)
Comment (v-rob) [reply src/script/lua_api/l_object.cpp:552]: I have to take responsibility here; I suggested it without realizing that these arguments were optional.

File: src/script/lua_api/l_object.cpp [hunk 14]
```diff
@@ -162,78 +164,67 @@ int ObjectRef::l_punch(lua_State *L)
...
-	float time_from_last_punch = 1000000;
-	if (lua_isnumber(L, 3))
-		time_from_last_punch = lua_tonumber(L, 3);
+
+	float time_from_last_punch = lua_isnil(L, 3) ?
+		1000000.0f : readParam<float>(L,3);
 	ToolCapabilities toolcap = read_tool_capabilities(L, 4);
-	dir.normalize();
+	v3f dir = lua_isnil(L, 5) ?
+		sao->getBasePosition() - puncher->getBasePosition() : check_v3f(L, 5);
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp:176]: https://github.com/minetest-mods/throwing_arrows/issues/6

`lua_isnoneornil` is perhaps a better choice here to ensure mod compatibility.

File: src/script/lua_api/l_object.cpp (hunk 13 above)
Comment (rubenwardy) [reply src/script/lua_api/l_object.cpp:2198]: (Missing type coercion)

File: src/script/lua_api/l_object.cpp [hunk 15]
```diff
@@ -647,17 +560,15 @@ int ObjectRef::l_get_bone_position(lua_State *L)
...
-	if (co == NULL)
+	ServerActiveObject *sao = getobject(ref);
+	if (sao == nullptr)
 		return 0;
-	// Do it
-	std::string bone;
-	if (!lua_isnil(L, 2))
-		bone = readParam<std::string>(L, 2);
+
+	std::string bone = readParam<std::string>(L, 2);
```
Comment (sfan5) [inline src/script/lua_api/l_object.cpp:567]: `bone` is also no longer optional (same in the function above)

File: src/script/lua_api/l_object.cpp [hunk 16]
```diff
@@ -1546,17 +1511,17 @@ int ObjectRef::l_hud_change(lua_State *L)
 	NO_MAP_LOCK_REQUIRED;
 	ObjectRef *ref = checkobject(L, 1);
 	RemotePlayer *player = getplayer(ref);
-	if (player == NULL)
+	if (player == nullptr)
 		return 0;
 
-	u32 id = lua_isnumber(L, 2) ? lua_tonumber(L, 2) : -1;
+	u32 id = luaL_checkint(L, 2);
```
Comment (sfan5) [inline src/script/lua_api/l_object.cpp:1517]: Assuming Krock's related comment is correct, the issue with `luaL_checkint` applies here too (strings are valid numbers, but the function won't accept them).

File: src/script/lua_api/l_object.cpp [hunk 17]
```diff
@@ -1570,15 +1535,16 @@ int ObjectRef::l_hud_get(lua_State *L)
 	NO_MAP_LOCK_REQUIRED;
 	ObjectRef *ref = checkobject(L, 1);
 	RemotePlayer *player = getplayer(ref);
-	if (player == NULL)
+	if (player == nullptr)
 		return 0;
 
-	u32 id = lua_tonumber(L, -1);
+	u32 id = luaL_checkint(L, 2);
```
Comment (sfan5) [inline src/script/lua_api/l_object.cpp:1541]: same

File: src/script/lua_api/l_object.cpp (hunk 15 above)
Comment (Zughy) [reply src/script/lua_api/l_object.cpp:567]: I don't think it should be, the function is literally called `get_bone_position`. Or is there something I'm missing @Jordach ? (If it is, there should be a comment somewhere)

File: src/script/lua_api/l_object.cpp (hunk 16 above)
Comment (Zughy) [reply src/script/lua_api/l_object.cpp:1517]: (assuming he's right) why exactly someone should pass a string instead of a number though? The API says nothing about it

Comment (rubenwardy) [reply src/script/lua_api/l_object.cpp:1517]: This is called type coercion. Lua should convert strings to numbers when used in a numeric context - although, the documents are unclear as to whether this is just with arithmetic or with other contexts too

Comment (Zughy) [reply src/script/lua_api/l_object.cpp:1517]: I mean, that's useful for things like `print("this is a number: " .. 55)` but I still don't see why someone should pass a string into this function (`hud_change`). Anyways, a solution could be to have a `readParam<int>(...)` checking for NaN, applying type coercion and then casting into an int (if it's a decimal).  
Also, the previous version was wrong too, because I could have passed a float, and HUDs take no floats

File: src/script/lua_api/l_object.cpp (hunk 15 above)
Comment (sfan5) [reply src/script/lua_api/l_object.cpp:567]: `""` is a valid way to refer to the root bone and `get_bone_position()` was a shortcut to achieve that previously.
It should continue to work.

File: src/script/lua_api/l_object.cpp (hunk 16 above)
Comment (sfan5) [reply src/script/lua_api/l_object.cpp:1517]: I don't know how important this really is (do other functions consistently take strings as numbers?), especially since you're not meant to do funny things to HUD IDs.
